src/datachain/lib/signal_schema.py

```diff
@@ -386,11 +386,20 @@ def row_to_features(
             else:
                 json, pos = unflatten_to_json_pos(fr, row, pos)  # type: ignore[union-attr]
                 obj = fr(**json)
-                if isinstance(obj, File):
-                    obj._set_stream(catalog, caching_enabled=cache)
+                SignalSchema._set_file_stream(obj, catalog, cache)
                 res.append(obj)
         return res
 
+    @staticmethod
+    def _set_file_stream(
+        obj: BaseModel, catalog: "Catalog", cache: bool = False
+    ) -> None:
+        if isinstance(obj, File):
+            obj._set_stream(catalog, caching_enabled=cache)
+        for field, finfo in obj.model_fields.items():
+            if ModelStore.is_pydantic(finfo.annotation):
+                SignalSchema._set_file_stream(getattr(obj, field), catalog, cache)
+
```

Comment:
This is a big fix that's needed to read files from wds regardless of the other changes in this PR. Right now, reading the wds files fails like this:

```python
>>> from datachain import C, DataChain
>>> from datachain.lib.webdataset import process_webdataset
>>> from datachain.lib.webdataset_laion import WDSLaion
>>> chain = DataChain.from_storage("gs://datachain-demo/datacomp-small/shards/00000000.tar", settings={"cache": True}).gen(laion=process_webdataset(spec=WDSLaion))
Processed: 1 rows [00:00, 1260.69 rows/s]
Generated: 2 rows [00:00, 2844.56 rows/s]
Cleanup: 1 tables [00:00, 5745.62 tables/s]
>>> row = next(chain.collect())
Processed: 1 rows [00:02,  2.66s/ rows]
Generated: 8953 rows [00:01, 8534.75 rows/s]
Cleanup: 1 tables [00:00, 39.20 tables/s]
>>> row[0].file.read()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/dave/Code/datachain/src/datachain/lib/file.py", line 213, in read
    with self.open() as stream:
  File "/Users/dave/micromamba/envs/datachain-py311/lib/python3.11/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/Users/dave/Code/datachain/src/datachain/lib/file.py", line 198, in open
    with VFileRegistry.resolve(self, self.location) as f:  # type: ignore[arg-type]
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dave/Code/datachain/src/datachain/lib/file.py", line 113, in resolve
    return reader.open(file, location)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dave/Code/datachain/src/datachain/lib/file.py", line 89, in open
    client = file._catalog.get_client(tar_file_uid.storage)
             ^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'get_client'
```

Reply:
Extracted to #443